epochStart/metachain/rewardsV2.go

```diff
@@ -0,0 +1,496 @@
+			rwdInfo.protocolRewards.Add(rwdInfo.protocolRewards, rewardsPerNode[shardID][i])
+		}
+	}
+
+	return rwdAddrValidatorInfo, accumulatedUnassigned
+}
+
+// IsInterfaceNil return true if underlying object is nil
+func (rc *rewardsCreatorV2) IsInterfaceNil() bool {
+	return rc == nil
+}
+
+func (rc *rewardsCreatorV2) prepareStakingData(eligibleNodesKeys map[uint32][][]byte) error {
+	sw := core.NewStopWatch()
+	sw.Start("prepareRewardsFromStakingSC")
+	defer func() {
+		sw.Stop("prepareRewardsFromStakingSC")
+		log.Debug("rewardsCreator.prepareRewardsFromStakingSC time measurements", sw.GetMeasurements())
+	}()
+
+	return rc.stakingDataProvider.PrepareStakingData(eligibleNodesKeys)
+}
+
+func (rc *rewardsCreatorV2) getEligibleNodesKeyMap(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+) map[uint32][][]byte {
+
+	eligibleNodesKeys := make(map[uint32][][]byte)
+	for shardID, validatorsInfoSlice := range validatorsInfo {
+		eligibleNodesKeys[shardID] = make([][]byte, len(validatorsInfoSlice))
+		for i, validatorInfo := range validatorsInfoSlice {
+			eligibleNodesKeys[shardID][i] = validatorInfo.PublicKey
+		}
+	}
+
+	return eligibleNodesKeys
+}
+
+func (rc *rewardsCreatorV2) computeRewardsPerNode(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+) (map[uint32][]*big.Int, *big.Int) {
+
+	// totalTopUpEligible is the cumulative top-up stake value for eligible nodes
+	totalTopUpEligible := rc.stakingDataProvider.GetTotalTopUpStakeEligibleNodes()
+	remainingToBeDistributed := rc.economicsDataProvider.RewardsToBeDistributedForBlocks()
+	topUpRewards := rc.computeTopUpRewards(remainingToBeDistributed, totalTopUpEligible)
+	baseRewards := big.NewInt(0).Sub(remainingToBeDistributed, topUpRewards)
+	nbBlocks := big.NewInt(int64(rc.economicsDataProvider.NumberOfBlocks()))
+	baseRewardsPerBlock := big.NewInt(0).Div(baseRewards, nbBlocks)
+	rc.fillBaseRewardsPerBlockPerNode(baseRewardsPerBlock)
+
+	accumulatedDust := big.NewInt(0)
+	baseRewardsPerNode, dust := rc.computeBaseRewardsPerNode(validatorsInfo, baseRewards)
+	accumulatedDust.Add(accumulatedDust, dust)
+	topUpRewardsPerNode, dust := rc.computeTopUpRewardsPerNode(topUpRewards, validatorsInfo)
+	accumulatedDust.Add(accumulatedDust, dust)
+
+	return aggregateBaseAndTopUpRewardsPerNode(baseRewardsPerNode, topUpRewardsPerNode), accumulatedDust
+}
+
+func (rc *rewardsCreatorV2) computeBaseRewardsPerNode(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+	baseRewards *big.Int,
+) (map[uint32][]*big.Int, *big.Int) {
+	baseRewardsPerNode := make(map[uint32][]*big.Int)
+	accumulatedRewards := big.NewInt(0)
+
+	for shardID, valInfoList := range validatorsInfo {
+		baseRewardsPerNode[shardID] = make([]*big.Int, len(valInfoList))
+		for i, valInfo := range valInfoList {
+			baseRewardsPerNode[shardID][i] = big.NewInt(0).Mul(
+				rc.mapBaseRewardsPerBlockPerValidator[shardID],
+				big.NewInt(0).SetInt64(int64(valInfo.NumSelectedInSuccessBlocks)))
```

Comment:
big.NewInt(0).SetUint - or use directly in big.New(int64(valInfo.NumSelectedInSuccessBlocks))

Reply:
done